lib/ansible/modules/cloud/amazon/ec2_eip.py

```diff
@@ -123,151 +128,183 @@
-    if not address_is_associated_with_device(ec2, address, device_id, isinstance):
-        return {'changed': False}
 
+def disassociate_ip_and_device(ec2, address, device_id, check_mode, is_aninstance=True):
     # If we're in check mode, nothing else to do
     if not check_mode:
-        if address.domain == 'vpc':
-            res = ec2.disassociate_address(
-                association_id=address.association_id)
+        if address.get('Domain') == 'vpc':
+            response = ec2.disassociate_address(
+                AssociationId=address.get('AssociationId'))
         else:
-            res = ec2.disassociate_address(public_ip=address.public_ip)
-
-        if not res:
-            raise EIPException('disassociation failed')
+            response = ec2.disassociate_address(
+                PublicIp=address.get('PublicIp'))
 
-    return {'changed': True}
+    changed = True
+    return response, changed
 
 
 def _find_address_by_ip(ec2, public_ip):
-    try:
-        return ec2.get_all_addresses([public_ip])[0]
-    except boto.exception.EC2ResponseError as e:
-        if "Address '{0}' not found.".format(public_ip) not in e.message:
-            raise
+    addresses = ec2.describe_addresses(PublicIps=[public_ip])
+    address = next(iter(addresses.values()))
+    if address:
+        return address[0]
 
 
-def _find_address_by_device_id(ec2, device_id, isinstance=True):
-    if isinstance:
-        addresses = ec2.get_all_addresses(None, {'instance-id': device_id})
+def _find_address_by_device_id(ec2, device_id, is_aninstance=True):
+    if is_aninstance:
+        device_filter = {'Name': 'instance-id', 'Values': [device_id]}
+        addresses = ec2.describe_addresses(Filters=[device_filter])
     else:
-        addresses = ec2.get_all_addresses(None, {'network-interface-id': device_id})
-    if addresses:
-        return addresses[0]
+        network_filter = {
+            'Name': 'network-interface-id', 'Values': [device_id]}
+        addresses = ec2.describe_addresses(Filters=[network_filter])
+
+    address = next(iter(addresses.values()))
+    if address:
+        return address[0]
 
 
-def find_address(ec2, public_ip, device_id, isinstance=True):
-    """ Find an existing Elastic IP address """
+def find_address(ec2, public_ip, device_id, is_aninstance=True):
+    # Find an existing Elastic IP address
     if public_ip:
         return _find_address_by_ip(ec2, public_ip)
-    elif device_id and isinstance:
+    elif device_id and is_aninstance:
         return _find_address_by_device_id(ec2, device_id)
     elif device_id:
-        return _find_address_by_device_id(ec2, device_id, isinstance=False)
+        return _find_address_by_device_id(ec2, device_id, is_aninstance=False)
+
+
+def address_is_associated_with_device(ec2, address, device_id, module, is_aninstance=True):
+    # Check if the elastic IP is currently associated with the device
+
+    if device_id.startswith('eni-'):
+        is_aninstance = False
```

Comment:
This is a pretty confusing place to be setting is_aninstance.  is_aninstance is also getting defaulted in many different places and passed around.  That makes it pretty hard for a developer to figure out what it should be at any given point.

Maybe another way of doing it would be to create an is_aninstance function that takes device_id as a parameter, then call that whenever you need to know if something is an instance.  Then you don't need to pass it around any more.